File: backend/internal/domain/transaction.go
package domain

import (
    "time"
    "github.com/google/uuid"
)

type TransactionType string

const (
    Income  TransactionType = "income"
    Expense TransactionType = "expense"
)

type Category struct {
    ID   uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
    Name string    `gorm:"not null" json:"name"`
    Icon string    `json:"icon"`
}

type Transaction struct {
    ID          uuid.UUID       `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
    UserID      uuid.UUID       `gorm:"type:uuid;not null" json:"user_id"`
    User        User            `json:"-"`
    CategoryID  uuid.UUID       `gorm:"type:uuid;not null" json:"category_id"`
    Category    Category        `json:"category"`
    Type        TransactionType `gorm:"type:varchar(10);not null" json:"type"`
    Amount      float64         `gorm:"not null" json:"amount"`
    Description string          `json:"description"`
    Date        time.Time       `gorm:"not null;default:now()" json:"date"`
    CreatedAt   time.Time       `json:"created_at"`
    UpdatedAt   time.Time       `json:"updated_at"`
}

File: frontend/src/types/transaction.ts
export type TransactionType = "income" | "expense";

export interface Category {
  id: string;
  name: string;
  icon: string;
}

export type TransactionType = "income" | "expense";

export interface Category {
  id: string;
  name: string;
  icon: string;
}

export interface Transaction {
  id: string;
  user_id: string;
  category_id: string;
  category: Category;
  type: TransactionType;
  amount: number;
  description: string;
  date: string | null;
  created_at: string | null;
  updated_at: string | null;
}


export interface CreateTransactionInput {
  category_id: string;
  type: TransactionType;
  amount: number;
  description?: string;
  date: string;
}

export interface UpdateTransactionInput {
  category_id?: string;
  type?: TransactionType;
  amount?: number;
  description?: string;
  date?: string;
}

export interface TransactionSummary {
  income: number;
  expense: number;
  balance: number;
  month: number;
  year: number;
}

export interface TransactionFilter {
  type?: string;
  category_id?: string;
  search?: string;
  start_date?: string;
  end_date?: string;
}

export interface Budget {
  id: string;
  category_id: string;
  category: Category;
  amount: number;
  month: number;
  year: number;
  spent: number;
  remaining: number;
  is_over: boolean;
}

export interface UpsertBudgetInput {
  category_id: string;
  amount: number;
  month: number;
  year: number;
}
